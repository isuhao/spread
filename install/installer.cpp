#include "installer.hpp"
#include "build_actions.hpp"
#include "rules/urlrule.hpp"

using namespace Spread;

Installer::Installer(Cache::Cache &_cache, RuleSet &_rules,
                     const std::string &_prefix)
  : cache(_cache), rules(_rules)
{
  build.reset(new ActionBuilder(cache, rules, _prefix));
}

void Installer::addHint(const Hash &hint)
{ build->addHint(hint); }
void Installer::addDep(const std::string &file, const Hash &hash)
{ build->addDep(file,hash); }
void Installer::addDir(const Directory *dir)
{ build->addDir(dir); }
void Installer::addDir(const Hash &hash, bool alsoAsHint)
{ build->addDir(hash, alsoAsHint); }

void Installer::getActions(ActionMap &acts)
{ build->build(acts); }

void Installer::addToCache(const Hash &h, const std::string &file)
{ cache.index.addFile(file, h); }

std::string Installer::getTmpFile(const Hash &h)
{ return cache.createTmpFilename(h); }

std::string Installer::brokenURL(const Hash &hash, const std::string &url)
{
  rules.reportBrokenURL(hash, url);

  // Find a replacement
  const Rule *r = rules.findRule(hash);
  if(!r || r->type != RST_URL) return "";

  // Looks like we found one. Extract and return it.
  const std::string &newUrl = URLRule::get(r)->url;
  assert(newUrl != url);
  return newUrl;
}
