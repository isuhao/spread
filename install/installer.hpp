#ifndef __SPREAD_INSTALLER_HPP
#define __SPREAD_INSTALLER_HPP

#include "action_installer.hpp"
#include "cache/cache.hpp"
#include "rules/ruleset.hpp"

/* Main installer job.

   Add dependencies (output files) through addDep and the addDir()
   variants, then run the job like normal.
 */

namespace Spread
{
  struct ActionBuilder;
  struct Installer : ActionInstaller
  {
    Installer(Cache::Cache &_cache, RuleSet &_rules,
              const std::string &_prefix = "");

    /* Add archive hint. The hash can be either for the archive file
       itself or for the directory of its contents.

       Archive hints are used to preload archive directories so we can
       find the files inside. If you do not 'prime' the installer with
       the necessary archives before installation, no archives will be
       used.
    */
    void addHint(const Hash &hint);

    // Add a single dependency. The file path is taken to be relative
    // to 'prefix'.
    void addDep(const std::string &file, const Hash &hash);

    // Add a directory of dependencies. All paths are relative to
    // 'prefix'.
    void addDir(const Directory *dir);
    void addDir(DirectoryCPtr dir);

    /* Add a directory by hash. This assumes that either the directory
       is available as a loadable cached object somewhere, or that
       there is a matching archive rule for the directory.

       If alsoAsHint==true (default), then any archive rule that
       contains this directory is loaded, if it exists.

       If the directory object does not exist, but an archive rule is
       found, the directory is added as a "blind" unpack. That means
       that whatever the archive contains is used as the output
       directory.
     */
    void addDir(const Hash &hash, bool alsoAsHint = true);

  private:
    // Implemented from ActionInstaller:
    std::string brokenURL(const Hash &hash, const std::string &url);
    void getActions(ActionMap &acts);
    std::string getTmpFile(const Hash &h);
    void addToCache(const Hash::DirMap &list);

    Cache::Cache &cache;
    RuleSet &rules;
    boost::shared_ptr<ActionBuilder> build;
  };
}

#endif
