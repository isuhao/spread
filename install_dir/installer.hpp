#ifndef __SPREAD_I_INSTALLER_HPP__
#define __SPREAD_I_INSTALLER_HPP__

#include <hash/hash.hpp>
#include <boost/shared_ptr.hpp>

namespace Spread
{
  struct Installer
  {
    /* Add a single file to the install list. The file path is taken
       to be relative to 'basedir'.

       If the file already exists and does not match a file added with
       remFile/remDir, the job will ask through owner.askWait()
       whether the user wants to overwrite the file.
    */
    virtual void addFile(const std::string &file, const Hash &hash) = 0;

    /* List a single file as already existing in the output directory.
       This has several uses:

       - if the file is both added with addFile() and remFile() with
         the SAME hash, then the file IGNORED in the install process
         (it's irrelevant whether or not any physical file matches the
         given hash - the location is not inspected.)

         Rationale: Files that are not scheduled to change in an
         update should not be touched, since that might inadvertently
         overwrite user changes.

       - if remFile() is called and NO matching file is listed with
         addFile(), then the file is DELETED. However if it does not
         match the requested hash, the user is asked first through
         owner.askWait().

         Rationale: Files removed in an output should be deleted, but
         user-modified files should not be removed without
         confirmation.

       - if a file is deleted in one places and added in another, the
         installer will usually rename/move the file as an
         optimization.

       The net effect is that if remDir is called with the known
       _existing_ contents of a directory, and addDir with the _new_
       (updated) contents of a directory, then the installer will
       replace or patch files as expected, create or delete files as
       required, and leave all unchanged files alone (even if they
       contain user modifications.) All affected files with user
       modifications cause a recoverable error.
     */
    virtual void remFile(const std::string &file, const Hash &hash) = 0;

    // Convenience functions:
    virtual void updateFile(const std::string &file, const Hash &oldH, const Hash &newH)
    { addFile(file, newH); remFile(file, oldH); }
    virtual void ignoreFile(const std::string &file)
    { Hash h("IGNORE"); updateFile(file, h, h); }

    /* Add/remove a directory of dependencies. All paths are relative
       to 'basedir'. You can also specify an optional path to prepend
       to all file paths. Paths will automatically be slash
       terminated. The result will be:

       prefix / path / path_in_directory

       These functions are equivalent of calling addFile()/remFile()
       multiple times in a row. If you want more functionality, see
       the hash version below.
    */
    virtual void addDir(const Hash::DirMap &dir, const std::string &path = "") = 0;
    virtual void remDir(const Hash::DirMap &dir, const std::string &path = "") = 0;

    /* 'Enhanced' hash versions of addDir()/remDir().

       The given hash may either refer to a directory object, or to an
       archive hash (in which case a matching archive rule must exist
       containing the dirhash.)

       In either case, if the directory matches a known archive
       dirhash, and the archive is available, it will be installed
       directly into the output directory. A pre-existing dir object
       is not required.

       Otherwise, the function REQUIRES that the directory is
       available as a binary loadable object somewhere (in the rule
       system or cache.)

       The installer may ask the RuleSet for any additional hints for
       finding the directory object, possibly fetching additional
       files and rules in the process.
     */
    virtual void addDir(const Hash &hash, const std::string &path = "") = 0;
    virtual void remDir(const Hash &hash, const std::string &path = "") = 0;

    /* Add a hint to the file-fetching system. A hint is a directory
       (or corresponding archive) hash that should be included in the
       file lookup process.

       Hints do not affect the list output files, only where we might
       FIND those output files.
     */
    virtual void addHint(const Hash &hint) = 0;
  };

  typedef boost::shared_ptr<Installer> InstallerPtr;
}
#endif
