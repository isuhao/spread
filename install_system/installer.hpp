#ifndef __SPREAD_INSTALLER_HPP
#define __SPREAD_INSTALLER_HPP

#include "cache/cache.hpp"
#include "rules/ruleset.hpp"
#include "job/job.hpp"

/* Main installer job.

   Add dependencies (output files) through addDep() and the addDir()
   variants, then run the job like normal.
 */

namespace Spread
{
  struct Installer : Job
  {
    Installer(Cache::Cache &_cache, RuleSet &_rules,
              const std::string &_prefix = "");

    /* Add a single file to the install list. The file path is taken
       to be relative to 'prefix'.

       If the file already exists and does not match a file added with
       remFile/remDir, then the execution may return a recoverable
       error, asking the user whether to keep or overwrite the file.
    */
    void addFile(const std::string &file, const Hash &hash);

    // Convenience functions
    void addFile(const std::string &file, const Hash &hash, const Hash &existing)
    { addFile(file, hash); remFile(file, existing); }
    void ignoreFile(const std::string &file, const Hash &hash)
    { addFile(file, hash, hash); }

    /* List a single file as already existing in the output directory.
       This has several uses:

       - if the file is both added with addFile() and listed with
         remFile() with the SAME hash, then the file IGNORED in the
         install process (it's irrelevant whether or not any physical
         file matches the given hash.) Use this to mark eg. initial
         config files when updating, so that the update doesn't
         inadvertently overwrite user changes.

       - if remFile() is called and NO matching file is listed with
         addFile(), then the file is DELETED.

       - if a file is deleted in one places and added in another, the
         installer will usually just rename/move the file.

       The net effect is that if remFile/remDir is called with the
       known _existing_ contents of a directory, and addFile/addDir
       with the _new_ (updated) contents of a directory, then the
       installer will replace or patch files as expected, add and
       delete files as required, and leave all unchanged files alone
       (even if they have user modifications.)

       The function is not necessary for patching. If patching an
       existing file (from the same location or not) can be used to
       create the destination file, then patching is used regardless
       of whether remFile() was called.
     */
    void remFile(const std::string &file, const Hash &hash);

    /* Add a directory of dependencies. All paths are relative to
       'prefix'. You can also specify an optional path to prepend to
       all file paths. The path MUST BE SLASH TERMINATED if you want
       it to act as a directory. The final path becomes:

       prefix + path + path_in_directory

       The given hash may either refer to a directory object, or to an
       archive hash (in which case the entire archive is added as-is.)

       The function REQUIRES that the directory is either available as
       a loadable object, or that there is a matching archive rule for
       the directory.

       The installer will ask the RuleSet for any additional hints for
       finding the directory object, required archive file(s), or any
       other rule that may help the installation.
     */
    void addDir(const Hash &hash, const std::string &path = "");

    /* The remFile()-based equivalent of remDir().

       Notes: Like addDir() this asks the rule system for extra hints,
       but unlike addDir() the hints are used for PATCHING the listed
       files (if possible), not for installing them.
     */
    void remDir(const Hash &hash, const std::string &path = "");

  private:
    struct _Internal;
    boost::shared_ptr<_Internal> ptr;
  };
}

#endif
