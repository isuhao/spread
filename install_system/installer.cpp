#include "installer.hpp"

using namespace Spread;

struct Installer::_Internal
{
  Internal(Cache::Cache &_cache, RuleSet &_rules, const std::string &_prefix)
  {}
};

Installer::Installer(Cache::Cache &cache, RuleSet &rules, const std::string &prefix)
{ ptr.reset(new _Internal(cache, rules, prefix)); }

void Installer::addFile(const std::string &file, const Hash &hash)
{
}

void Installer::remFile(const std::string &file, const Hash &hash)
{
}

void Installer::addDir(const Hash &hash, const std::string &path)
{
}

void Installer::remDir(const Hash &hash, const std::string &path)
{
}
