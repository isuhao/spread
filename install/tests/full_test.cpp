#include "installer.hpp"
#include <iostream>

using namespace std;
using namespace Spread;

Hash hello("hello", 5);
Hash world("world", 5);
Hash testsh("ctEjJBRstghw4_UpmjBdhwJZFl8faISyIeEk2sOH5LLfAQ");
Hash zipfile("bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ");
Hash zipzip("RKEFhXsBxTYXmQt4oMxU_RQk_t7F0Oj8RMGALSJKh6UfAg");
Hash testzip("HjZUnmXcFYkS8mTsovU-oVKEdKoGbF7Ga3NlNWYX6spMAQ");
DirectoryPtr testDir, zipDir, zipzipDir;

Hash testDirH, zipDirH, zipzipDirH;

void setupDirs()
{
  testDir.reset(new Directory);
  zipDir.reset(new Directory);
  zipzipDir.reset(new Directory);

  testDir->dir["hello.out"] = hello;
  testDir->dir["dir/world.out"] = world;
  zipDir->dir["test.sh"] = testsh;
  zipzipDir->dir["testsh.zip"] = zipfile;

  using namespace Mangle::Stream;

  testDirH = testDir->write("_test.dir");
  zipDirH = zipDir->write("_zip.dir");
  zipzipDirH = zipzipDir->write("_zipzip.dir");
}

Cache::Cache cache;
RuleSet rules;

#define ADD(f,h) assert(cache.index.addFile(f)==(h))

void setup()
{
  setupDirs();

  cache.tmpDir = "_full_tmp/";

  ADD("_test.dir", testDirH);
  ADD("_zip.dir", zipDirH);
  ADD("_zipzip.dir", zipzipDirH);
  ADD("testdir.zip",testzip);
  ADD("zipzip.zip",zipzip);

  rules.addArchive(zipfile, zipDirH, "ZIP1");
  rules.addArchive(testzip, testDirH, "ZIP TEST");
  rules.addArchive(zipzip, zipzipDirH, "ZIPZIP");
}

void test(ActionInstaller &a, const std::string &msg)
{
  cout << "\n" << msg << "\n";
  JobInfoPtr info = a.run();
  assert(info->isFinished());
  if(info->isError()) cout << "ERROR";
  else if(info->isSuccess()) cout << "DONE";
  else cout << "UNKNOWN STATUS";
  cout << ": status=" << info->getMessage() << "\nprogress=" << info->getCurrent()
       << "/" << info->getTotal() << endl;
}

struct AdHoc
{
  ActionInstaller &act;
  std::string prefix, msg;

  AdHoc(ActionInstaller &a, const std::string& pref, const std::string &m)
    : act(a), prefix(pref), msg(m) {}

  ~AdHoc() { test(act, msg + " => " + prefix + "/"); }
};

#define TEST(msg,path) std::string _prefix="_full/"+std::string(path);Installer inst(cache, rules, _prefix);AdHoc _ad(inst,_prefix,(msg))

int main()
{
  setup();

  {
    TEST("Empty installer","t1");
  }

  {
    TEST("Requesting non-existing file","t2");
    inst.addDep("output.dat", hello);
  }

  {
    TEST("Requesting existing file","t3");
    inst.addDep("test.dir", testDirH);
  }

  {
    TEST("File from zip (no hint)", "t4");
    inst.addDep("test.zip", zipfile);
  }

  {
    TEST("File from zip (with hint)", "t5");
    inst.addDep("test.zip", zipfile);
    inst.addHint(zipzip);
  }

  {
    TEST("Repeating with no hint (using cache)", "t6");
    inst.addDep("test.zip", zipfile);
  }

  {
    TEST("Unpack two zips into one dir", "t7");
    inst.addDir(zipDirH);
    inst.addDir(zipzipDirH);
    inst.addDir(testDirH);
  }

  {
    TEST("Repeating previously non-existing file","t8");
    inst.addDep("output.dat", hello);
  }
  return 0;
}
