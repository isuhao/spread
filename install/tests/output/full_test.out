
Empty installer => _full/t1/
DONE: status=Installing
progress=0/0

Requesting non-existing file => _full/t2/
ERROR: status=Don't know how to satisfy the following dependencies:
  LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
    _full/t2/output.dat
progress=0/0

Requesting existing file => _full/t3/
DONE: status=Installing
progress=0/0

File from zip (no hint) => _full/t4/
ERROR: status=Don't know how to satisfy the following dependencies:
  bJUN2VkwS-cRpIlPRil-yWIxuoOfiWUG4li9IxH5MGWkAQ
    _full/t4/test.zip
progress=0/0

File from zip (with hint) => _full/t5/
DONE: status=Installing
progress=0/0

Repeating with no hint (using cache) => _full/t6/
DONE: status=Installing
progress=0/0

Unpack two zips into one dir => _full/t7/
DONE: status=Installing
progress=0/0

Repeating previously non-existing file => _full/t8/
DONE: status=Installing
progress=0/0
